### PR TITLE
osd/OSD: choose heartbeat peers more carefully

### PR DESCRIPTION
**Pull request by xiexingguo**
By default, monitor requires at least two valid failure votes/reports from
different hosts to mark an OSD down, which turns out to be impossible sometimes
for a replicated-pool of size of 2 in those clusters made up of hosts
with contiguous labeled OSDs.

This patch instead does a breadth-first search based on the highest level of
failure domain at cluster-wide, to try to make heartbeat peers can cover all failure domains
whenever possible, which can hopefully help accelerating osd failure detection
in the above case..

### CODE REVIEW

File: src/crush/CrushWrapper.cc [hunk 1]
```diff
@@ -1524,6 +1524,17 @@ int CrushWrapper::get_parent_of_type(int item, int type, int rule) const
   return 0; // not found
 }
 
+void CrushWrapper::get_subtree_of_type(int type, vector<int> *subtrees) const
+{
+  for (int i = 0; i < crush->max_buckets; ++i) {
```
Comment (liewegas) [inline src/crush/CrushWrapper.cc]: this should only do this search from the roots (use get_roots()) or else there is a ton of duplicated work doing recursive descent from every bucket

**Comment by liewegas** [comment]
This looks great except for the way get_subtree_of_type is called

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -4317,7 +4317,9 @@ void OSD::maybe_update_heartbeat_peers()
 	dout(10) << "maybe_update_heartbeat_peers forcing update after " << dur << " seconds" << dendl;
 	heartbeat_set_peers_need_update();
 	last_heartbeat_resample = now;
-	reset_heartbeat_peers();   // we want *new* peers!
+        if (is_waiting_for_healthy()) {
+	  reset_heartbeat_peers();   // we want *new* peers!
```
Comment (Yan-waller) [inline src/osd/OSD.cc]: nit, indent

Comment (xiexingguo) [reply src/osd/OSD.cc]: it's a space vs table issue.

**Comment by Yan-waller** [comment]
http://pulpito.ceph.com/xxg-2018-08-09_12:35:56-rados-wip-yanjun-testing-distro-basic-smithi/

**Comment by xiexingguo** [comment]
@Yan-waller Thanks!

**Pull request merged**